src/client/java/daos-java-api/src/main/native/com_intel_daos_client_DaosFsClient.c

```diff
@@ -0,0 +1,911 @@
+	src_dir_path = strdup(src_path);
+	src_base_path = strdup(src_path);
+	dest_dir_path = strdup(dest_path);
+	dest_base_path = strdup(dest_path);
+	char * src_dir = dirname(src_dir_path);
+	char * src_base = basename(src_base_path);
+	char * dest_dir = dirname(dest_dir_path);
+	char * dest_base = basename(dest_base_path);
+	dfs_obj_t *src_dir_handle = NULL, *dest_dir_handle = NULL;
+	mode_t tmp_mode;
+	dfs_t *dfs = *(dfs_t**)&dfsPtr;
+	int rc = dfs_lookup(dfs, src_dir, O_RDWR, &src_dir_handle, &tmp_mode, NULL);
+	if (rc) {
+		char *tmp = "Cannot open source directory (%s)";
+		char *msg = (char *)malloc(strlen(tmp) + strlen(src_dir));
+		sprintf(msg, tmp, src_dir);
+		throw_exception(env, msg, rc);
+		goto out;
+	}
+	if (strcmp(src_dir, dest_dir) == 0) {
+		dest_dir_handle = src_dir_handle;
+	} else {
+		rc = dfs_lookup(dfs, dest_dir, O_RDWR, &dest_dir_handle, &tmp_mode, NULL);
+		if (rc) {
+			char *tmp = "Cannot open destination directory (%s)";
+			char *msg = (char *)malloc(strlen(tmp) + strlen(dest_dir));
+			sprintf(msg, tmp, dest_dir);
+			throw_exception(env, msg, rc);
+			goto out;
+		}
+	}
+	rc = dfs_move(dfs, src_dir_handle, src_base, dest_dir_handle, dest_base, NULL);
+	if (rc) {
+		char *tmp = "Failed to move source path (%s) to destination path (%s)";
+		char *msg = (char *)malloc(strlen(tmp) + strlen(src_path) + strlen(dest_path));
+		sprintf(msg, tmp, src_path, dest_path);
+		throw_exception(env, msg, rc);
+	}
+out:
+	if (src_dir_path) free(src_dir_path);
+	if (src_base_path) free(src_base_path);
+	if (dest_dir_path) free(dest_dir_path);
+	if (dest_base_path) free(dest_base_path);
+	if (src_dir_handle) dfs_release(src_dir_handle);
+	if (src_dir_handle != dest_dir_handle) dfs_release(dest_dir_handle);
+	(*env)->ReleaseStringUTFChars(env, srcPath, src_path);
+	(*env)->ReleaseStringUTFChars(env, destPath, dest_path);
+}
+
+static int mkdirs(dfs_t *dfs, char *path, int mode, unsigned char recursive, dfs_obj_t **handle, char *msg){
+	char *dirs = NULL, *bases=NULL, *dir=NULL, *base=NULL;
+	dfs_obj_t *parent_handle = NULL;
+	mode_t parent_mode;
+	int rc = dfs_lookup(dfs, path, O_RDWR, handle, &parent_mode, NULL);
+	if (rc == -DER_NONEXIST || rc == -ENOENT || rc == ENOENT) {
+		if(recursive){
+			// recursively create it
+			dirs = strdup(path);
+			dir = dirname(dirs);
+			bases = strdup(path);
+			base = basename(bases);
+			rc = mkdirs(dfs, dir, mode, recursive, &parent_handle, msg);
+			if (rc)	{
+				if (msg[0] == '\0') {
+					int dir_len = strlen(dir);
+					int max_len = ERROR_PATH_LEN - 4;
+					int l = dir_len>max_len ? max_len:dir_len;
+					memcpy(msg, dir, l);
+					if (l == max_len) {
+						char *suffix = "...\0";
+						memcpy(msg+l, suffix, 4);
+					} else {
+						msg[l] = '\0';
+					}
+				}
+				goto out;
+			}
+			rc = dfs_mkdir(dfs, parent_handle, base, mode);
+			if (rc == ERROR_NOT_EXIST) { // code to mitigate DAOS concurrency issue
```

Comment:
im not sure what concurrency issue this is.
but if mkdir fails with no exist, you should just fail the API operation. mkdir is not an open!

Reply:
It's workaround to move on our integration test with Spark. the concurrency issue will be fixed by your conditional update feature.